lib/CGameInfoCallback.cpp

```diff
@@ -520,6 +520,28 @@ EBuildingState::EBuildingState CGameInfoCallback::canBuildStructure( const CGTow
 	if(vstd::contains(t->forbiddenBuildings, ID))
 		return EBuildingState::FORBIDDEN; //forbidden
 
+	std::function<bool(BuildingID id)> allowedTest;
+	std::function<bool(BuildingID id)> possiblyNotBuiltTest;
+
+	allowedTest = [&](BuildingID id) -> bool
+	{
+		if (vstd::contains(t->forbiddenBuildings, id))
+		{
+			return false;
+		}
+
+		return t->genBuildingRequirements(id, true).satisfiable(allowedTest, possiblyNotBuiltTest);
+	};
+
+	possiblyNotBuiltTest = [&](BuildingID id) -> bool
+	{
+		//TODO consider destroing
```

Comment:
This is about building A which need building B not to be built; can be expressed by genBuildingRequirements(A) = NoneOf(B).

If VCMI allow to destroy building, possiblyNotBuiltTest should rather return (!t->hasBuilt(id) || t->canBeDestroyed(id)) or just (true) if any building can be destroyed.

Assuming that is ever planned.